cs1/tokenbucket.py

```diff
@@ -0,0 +1,29 @@
+import time
+
+
+class TokenBucket:
+
+    def __init__(self, tokens, time_unit):
+        self.tokens = tokens
+        self.time_unit = time_unit
+        self.bucket = tokens
+        self.last_check = time.time()
+
+    def consume(self):
+        if (self.bucket < 1):
+            time.wait(self.time_unit / self.tokens)
```

Comment:
The delay can be more aggressive, since if this limit is hit, user code is breaking the UI.